File: lwaf_math/PolynomialCoefficientSolver.java
package lwaf_math;

public class PolynomialCoefficientSolver {
    private final int order;
    private final float[][] matrix;
    private final float[] composition;

    public PolynomialCoefficientSolver(int order) {
        this.order = order;
        this.matrix = new float[order][order];
        this.composition = new float[order];
    }

    public void setRow(int row, float ...value) {
        assert value.length == order;
        matrix[row] = value;
    }

    public void setValues(float ...values) {
        assert values.length == order;
        System.arraycopy(values, 0, composition, 0, values.length);
    }

    public float[] solve() {
        float[] result = new float[order];

        for (int i = 0; i < order; ++i) {
            for (int j = i + 1; j < order; ++j) {
                subtractRow(j, i, i);
            }
        }

        for (int i = order - 2; i >= 0; --i) {
            for (int j = i + 1; j < order; ++j) {
                subtractRow(i, j, j);
            }
        }

        for (int i = 0; i < order; ++i) {
            result[i] = composition[i] / matrix[i][i];
        }

        return result;
    }

    private void subtractRow(int row, int sourceRow, int elementToZero) {
        float multiple = matrix[row][elementToZero] / matrix[sourceRow][elementToZero];
        subMultiple(row, sourceRow, multiple);
    }

    private void subMultiple(int row, int sourceRow, float multiple) {
        for (int i = 0; i < order; ++i) {
            matrix[row][i] -= matrix[sourceRow][i] * multiple;
        }

        composition[row] -= composition[sourceRow] * multiple;
    }
}

File: src/lwaf_util/PolynomialCoefficientSolver.kt
package lwaf_util

class PolynomialCoefficientSolver(private val order: Int) {
    private val matrix: Array<FloatArray> = Array(order) { FloatArray(order) }
    private val composition: FloatArray = FloatArray(order)

    fun setRow(row: Int, vararg value: Float) {
        assert(value.size == order)
        matrix[row] = value
    }

    fun setValues(vararg values: Float) {
        assert(values.size == order)
        System.arraycopy(values, 0, composition, 0, values.size)
    }

    fun solve(): FloatArray {
        val result = FloatArray(order)

        for (i in 0 until order) {
            for (j in i + 1 until order) {
                subtractRow(j, i, i)
            }
        }

        for (i in order - 2 downTo 0) {
            for (j in i + 1 until order) {
                subtractRow(i, j, j)
            }
        }

        for (i in 0 until order) {
            result[i] = composition[i] / matrix[i][i]
        }

        return result
    }

    private fun subtractRow(row: Int, sourceRow: Int, elementToZero: Int) {
        val multiple = matrix[row][elementToZero] / matrix[sourceRow][elementToZero]
        subMultiple(row, sourceRow, multiple)
    }

    private fun subMultiple(row: Int, sourceRow: Int, multiple: Float) {
        for (i in 0 until order) {
            matrix[row][i] -= matrix[sourceRow][i] * multiple
        }

        composition[row] -= composition[sourceRow] * multiple
    }
}
